### PR TITLE
#197. notification re-transmittion added. PDRs below 1 work

### LINKED ISSUE #197: Add PDR<1 and re-transmit mechanism
- clearly define communication protocol

### PR DESCRIPTION
**Pull request by Raa37**

### CODE REVIEW

File: DotBot.py [hunk 1]
```diff
@@ -5,6 +5,7 @@
 # local
 import SimEngine
 import Wireless
+import Orchestrator
```
Comment (twatteyne) [inline DotBot.py]: big big big no-no, DotBot cannot konw about Orchestrator

File: DotBot.py [hunk 2]
```diff
@@ -18,6 +19,8 @@ class DotBot(Wireless.WirelessDevice):
     A single DotBot.
     '''
 
+    COMM_DOWNSTREAM_PERIOD_S      = 1
```
Comment (twatteyne) [inline DotBot.py]: certainly not the right name, DotBot doesn't gnerate any downstream data...

File: DotBot.py [hunk 3]
```diff
@@ -36,6 +39,9 @@ def __init__(self, dotBotId, x, y, floorplan):
         # timestamps of when movement starts/stops
         self.tsMovementStart      = None
         self.tsMovementStop       = None
+        # sequence numbers (to filter out duplicate commands and notifications)
+        self.seqNumMovement       = None
```
Comment (twatteyne) [inline DotBot.py]: what does movement mean? Do you refer to commands (Orchestrator->DotBot) and notifications (DotBot->orchestrator)?

File: DotBot.py [hunk 4]
```diff
@@ -162,6 +176,13 @@ def _transmit(self):
             sender      = self,
         )
 
+        # schedule re-transmit
```
Comment (twatteyne) [inline DotBot.py]: "schedule retransmission"

File: DotBot.py [hunk 5]
```diff
@@ -162,6 +176,13 @@ def _transmit(self):
...
+        self.simEngine.schedule(
+            ts=self.simEngine.currentTime() + self.COMM_DOWNSTREAM_PERIOD_S,
```
Comment (twatteyne) [inline DotBot.py]: indent

**Review by twatteyne — CHANGES_REQUESTED** [review]
small changes, should be easy to fix. PLease add a comment indicating what you did for testing

**Comment by Raa37** [comment]
Testing method:
1. changed PDR to 0.8 and saw how non of the cells were placed correctly anymore.
2. added retransmission and sequence numbers and checked that simulation ran to completion, placing cells in correct positions.
3. I then ran simulation with PDR = 0.3 to check that time to completion increases yet still completes. and indeed it took half an hou8r more to complete (in simulation time)

File: DotBot.py [hunk 6]
```diff
@@ -40,7 +39,7 @@ def __init__(self, dotBotId, x, y, floorplan):
         self.tsMovementStart      = None
         self.tsMovementStop       = None
         # sequence numbers (to filter out duplicate commands and notifications)
-        self.seqNumMovement       = None
+        self.seqNumCommand       = None
```
Comment (twatteyne) [inline DotBot.py]: Indent

File: DotBot.py [hunk 7]
```diff
@@ -19,7 +18,7 @@ class DotBot(Wireless.WirelessDevice):
     A single DotBot.
     '''
 
-    COMM_DOWNSTREAM_PERIOD_S      = 1
+    RETRY_TRANSMIT_PERIOD_S      = 1
```
Comment (twatteyne) [inline DotBot.py]: RETRY_TIMEOUT_S

**Review by twatteyne — CHANGES_REQUESTED** [review]
Once fixed, please merge

**Pull request merged**